src/osd.c

```diff
@@ -223,6 +223,138 @@ preview_cycled_view(struct view *view)
 	wlr_scene_node_raise_to_top(osd_state->preview_node);
 }
 
+static void
+osd_draw(cairo_t *cairo, int w, int h, struct wl_list *node_list,
+	int current_view_index, struct theme *theme, bool show_workspace,
+	const char *workspace_name)
```

Comment:
Suggest changing to `render_osd(struct output *output)` and moving everything rendering related from `osd_update()` to the top / bottom of this new function. The `struct server *server` pointer can be retrieved via `output->server`.

Reply:
I've done something along these lines, but still have a separation between `display_osd` (which does the work to get OSD onscreen) and `render_osd` (which does not rely on the rest of the compositor for the most part, and simply fills in pixels, because there's plenty of complexity in just that). See how you like it.